gef.py

```diff
@@ -1443,95 +1677,70 @@ def hexdump(source, length=0x10, separator=".", show_raw=False, show_symbol=True
+        gdb.execute("set logging overwrite")
+        gdb.execute(f"set logging file {self.redirection_target_file}")
+        gdb.execute("set logging redirect on")
+        gdb.execute("set logging on")
+        return
 
+    def __exit__(self, *exc) -> None:
+        """Disable the output redirection, if any."""
+        gdb.execute("set logging off")
+        gdb.execute("set logging redirect off")
+        return
 
-def enable_redirect_output(to_file="/dev/null"):
+@deprecated("Use `RedirectOutputContext()` context manager")
+def enable_redirect_output(to_file: str = "/dev/null") -> None:
     """Redirect all GDB output to `to_file` parameter. By default, `to_file` redirects to `/dev/null`."""
     gdb.execute("set logging overwrite")
-    gdb.execute("set logging file {:s}".format(to_file))
+    gdb.execute(f"set logging file {to_file}")
     gdb.execute("set logging redirect on")
     gdb.execute("set logging on")
     return
 
-
-def disable_redirect_output():
+@deprecated("Use `RedirectOutputContext()` context manager")
+def disable_redirect_output() -> None:
     """Disable the output redirection, if any."""
     gdb.execute("set logging off")
     gdb.execute("set logging redirect off")
     return
 
 
-@lru_cache()
-def get_gef_setting(name):
-    """Read global gef settings.
-    Return None if not found. A valid config setting can never return None,
-    but False, 0 or ""."""
-    global __config__
-    setting = __config__.get(name, None)
-    if not setting:
-        return None
-    return setting[0]
-
-
-def set_gef_setting(name, value, _type=None, _desc=None):
-    """Set global gef settings.
-    Raise ValueError if `name` doesn't exist and `type` and `desc`
-    are not provided."""
-    global __config__
-
-    if name not in __config__:
-        # create new setting
-        if _type is None or _desc is None:
-            raise ValueError("Setting '{}' is undefined, need to provide type and description".format(name))
-        __config__[name] = [_type(value), _type, _desc]
-        return
-
-    # set existing setting
-    func = __config__[name][1]
-    __config__[name][0] = func(value)
-    get_gef_setting.cache_clear()
-    return
-
-
-def gef_makedirs(path, mode=0o755):
+def gef_makedirs(path: str, mode: int = 0o755) -> str:
     """Recursive mkdir() creation. If successful, return the absolute path of the directory created."""
-    abspath = os.path.expanduser(path)
-    abspath = os.path.realpath(abspath)
-
-    if os.path.isdir(abspath):
-        return abspath
-
-    os.makedirs(abspath, mode=mode, exist_ok=True)
-    return abspath
+    fpath = pathlib.Path(path)
+    if not fpath.is_dir():
+        fpath.mkdir(mode=mode, exist_ok=True, parents=True)
+    return fpath.absolute()
 
 
 @lru_cache()
-def gdb_lookup_symbol(sym):
+def gdb_lookup_symbol(sym: str):  # -> Optional[Tuple[Optional[str], Optional[Tuple[gdb.Symtab_and_line, ...]]]]
```

Comment:
```suggestion
def gdb_lookup_symbol(sym: str) -> Optional[Tuple["gdb.Symtab_and_line", ...]]:
```

somehow missed the index `[1]` at the end (which greatly simplifies the type hint) when I wrote this comment